File: minijinja/src/debug.rs
use std::fmt;

use crate::compiler::tokens::Span;
use crate::error::ErrorKind;
use crate::value::Value;

/// This is a snapshot of the debug information.
#[cfg_attr(docsrs, doc(cfg(feature = "debug")))]
#[derive(Default)]
pub(crate) struct DebugInfo {
    pub(crate) template_source: Option<String>,
    pub(crate) context: Option<Value>,
    pub(crate) referenced_names: Option<Vec<String>>,
}

struct VarPrinter<'x>(Value, &'x [String]);

impl<'x> fmt::Debug for VarPrinter<'x> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut m = f.debug_struct("Referenced variables:");
        let mut vars = self.1.to_owned();
        vars.sort();
        for var in &vars {
            match self.0.get_attr(var) {
                Ok(val) => m.field(var, &val),
                Err(_) => m.field(var, &Value::UNDEFINED),
            };
        }
        m.finish()
    }
}

impl DebugInfo {
    /// If available this contains a reference to the source string.
    pub fn source(&self) -> Option<&str> {
        self.template_source.as_deref()
    }

    /// Provides access to a snapshot of the context.
    ///
    /// The context is created at the time the error was created if that error
    /// happened during template rendering.
    pub fn context(&self) -> Option<Value> {
        self.context.clone()
    }

    /// Returns a narrowed down set of referenced names from the context
    /// where the error happened.
    ///
    /// This function is currently internal and only used for the default
    /// error printing.  This could be exposed but it's a highly specific
    /// API.
    pub(crate) fn referenced_names(&self) -> Option<&[String]> {
        self.referenced_names.as_deref()
    }
}

pub(super) fn render_debug_info(
    f: &mut fmt::Formatter,
    name: Option<&str>,
    kind: ErrorKind,
    line: Option<usize>,
    span: Option<Span>,
    info: &DebugInfo,
) -> fmt::Result {
    if let Some(source) = info.source() {
        let title = format!(
            " {} ",
            name.unwrap_or_default()
                .rsplit(&['/', '\\'])
                .next()
                .unwrap_or("Template Source")
        );
        writeln!(f)?;
        writeln!(f, "{:-^1$}", title, 79).unwrap();
        let lines: Vec<_> = source.lines().enumerate().collect();
        let idx = line.unwrap_or(1).saturating_sub(1);
        let skip = idx.saturating_sub(3);
        let pre = lines.iter().skip(skip).take(3.min(idx)).collect::<Vec<_>>();
        let post = lines.iter().skip(idx + 1).take(3).collect::<Vec<_>>();
        for (idx, line) in pre {
            writeln!(f, "{:>4} | {}", idx + 1, line).unwrap();
        }

        writeln!(f, "{:>4} > {}", idx + 1, lines[idx].1).unwrap();
        if let Some(span) = span {
            if span.start_line == span.end_line {
                writeln!(
                    f,
                    "     i {}{} {}",
                    " ".repeat(span.start_col),
                    "^".repeat(span.end_col - span.start_col),
                    kind,
                )?;
            }
        }

        for (idx, line) in post {
            writeln!(f, "{:>4} | {}", idx + 1, line).unwrap();
        }
        write!(f, "{:~^1$}", "", 79).unwrap();
    }
    if let Some(ctx) = info.context() {
        if let Some(vars) = info.referenced_names() {
            writeln!(f)?;
            writeln!(f, "{:#?}", VarPrinter(ctx, vars))?;
        }
        write!(f, "{:-^1$}", "", 79).unwrap();
    }
    Ok(())
}

File: minijinja-go/internal/errors/debug.go
package errors

import (
	goerrors "errors"
	"fmt"
	"sort"
	"strings"

	"github.com/mitsuhiko/minijinja/minijinja-go/v2/syntax"
	"github.com/mitsuhiko/minijinja/minijinja-go/v2/value"
)

// DebugInfo is a snapshot of debug information captured during rendering.
type DebugInfo struct {
	TemplateSource   string
	ReferencedLocals map[string]value.Value
}

func formatErrorWithDebug(f fmt.State, err *Error, includeChain bool) {
	_, _ = fmt.Fprint(f, err.Error())
	if err.DebugInfo != nil {
		renderDebugInfo(f, err)
	}

	if includeChain {
		for cause := goerrors.Unwrap(err); cause != nil; cause = goerrors.Unwrap(cause) {
			_, _ = fmt.Fprint(f, "\n\ncaused by: ")
			if next, ok := cause.(*Error); ok {
				formatErrorWithDebug(f, next, false)
			} else {
				_, _ = fmt.Fprintf(f, "%v", cause)
			}
		}
	}
}

func renderDebugInfo(f fmt.State, err *Error) {
	info := err.DebugInfo
	if info == nil {
		return
	}

	if info.TemplateSource != "" {
		title := fmt.Sprintf(" %s ", templateTitle(err.Name))
		_, _ = fmt.Fprint(f, "\n")
		_, _ = fmt.Fprintln(f, centerLine(title, '-', 79))

		lines := strings.Split(info.TemplateSource, "\n")
		lineIdx := 0
		if err.Span != nil && err.Span.StartLine > 0 {
			lineIdx = int(err.Span.StartLine - 1)
		}
		if lineIdx >= len(lines) {
			lineIdx = len(lines) - 1
		}
		if lineIdx < 0 {
			lineIdx = 0
		}

		skip := lineIdx - 3
		if skip < 0 {
			skip = 0
		}
		for idx := skip; idx < lineIdx && idx < len(lines); idx++ {
			_, _ = fmt.Fprintf(f, "%4d | %s\n", idx+1, lines[idx])
		}

		if lineIdx < len(lines) {
			_, _ = fmt.Fprintf(f, "%4d > %s\n", lineIdx+1, lines[lineIdx])
		}

		if err.Span != nil && err.Span.StartLine == err.Span.EndLine {
			_, _ = fmt.Fprintf(
				f,
				"     i %s%s %s\n",
				strings.Repeat(" ", int(err.Span.StartCol)),
				strings.Repeat("^", caretWidth(err.Span)),
				err.Kind,
			)
		}

		for idx := lineIdx + 1; idx <= lineIdx+3 && idx < len(lines); idx++ {
			_, _ = fmt.Fprintf(f, "%4d | %s\n", idx+1, lines[idx])
		}
		_, _ = fmt.Fprint(f, strings.Repeat("~", 79))
	}

	_, _ = fmt.Fprint(f, "\n")
	renderReferencedLocals(f, info.ReferencedLocals)
	_, _ = fmt.Fprint(f, strings.Repeat("-", 79))
}

func renderReferencedLocals(f fmt.State, locals map[string]value.Value) {
	if len(locals) == 0 {
		_, _ = fmt.Fprint(f, "No referenced variables\n")
		return
	}

	_, _ = fmt.Fprint(f, "Referenced variables:\n")
	keys := make([]string, 0, len(locals))
	for key := range locals {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		_, _ = fmt.Fprintf(f, "    %s: %s\n", key, locals[key].Repr())
	}
}

func caretWidth(span *syntax.Span) int {
	if span == nil {
		return 0
	}
	if span.EndCol <= span.StartCol {
		return 1
	}
	return int(span.EndCol - span.StartCol)
}

func templateTitle(name string) string {
	if name == "" {
		return "Template Source"
	}
	parts := strings.FieldsFunc(name, func(r rune) bool { return r == '/' || r == '\\' })
	if len(parts) == 0 {
		return "Template Source"
	}
	return parts[len(parts)-1]
}

func centerLine(title string, fill rune, width int) string {
	if len(title) >= width {
		return title
	}
	pad := width - len(title)
	left := pad / 2
	right := pad - left
	return strings.Repeat(string(fill), left) + title + strings.Repeat(string(fill), right)
}
